R-Py parity for ml-regression page

kvdesai opened the pull request.

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+<!-- #region -->
+This page shows how to use Plotly charts for displaying various types of regression models, starting from simple models like Linear Regression and progressively move towards models like Decision Tree and Polynomial Features. We highlight various capabilities of plotly, such as comparative analysis of the same model with different parameters, displaying Latex, and [surface plots](https://plotly.co ...
```
```suggestion
This page shows how to use Plotly charts for displaying various types of regression models, starting from simple models like [Linear Regression](https://parsnip.tidymodels.org/reference/linear_reg.html) and progressively move towards models like Decision Tree and Polynomial Features. We highlight various capabilities of plotly, such as comparative analysis of the same model with different parameters, displaying Latex, and [surface plots](https://plotly.com/r/3d-surface-plots/) for 3D data.
```

Can we also add the in-line hyperlinks to "Decision Tree" and "Polynomial Features" for parity? In the Python docs we have hyperlinks for the `sk-learn` docs, maybe we can have something similar with the `parsnip` or `tidymodels` references.

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+<!-- #region -->
+This page shows how to use Plotly charts for displaying various types of regression models, starting from simple models like Linear Regression and progressively move towards models like Decision Tree and Polynomial Features. We highlight various capabilities of plotly, such as comparative analysis of the same model with different parameters, displaying Latex, and [surface plots](https://plotly.co ...
+
+We will use [tidymodels](https://tidymodels.tidymodels.org/) to split and preprocess our data and train various regression models. Tidymodels is a popular Machine Learning (ML) library in R that is compatible with the "tidyverse" concepts, and offers various tools for creating and training ML algorithms, feature engineering, data cleaning, and evaluating and testing models. It is the next-gen ver ...
```
```suggestion
We will use [tidymodels](https://tidymodels.tidymodels.org/) to split and preprocess our data and train various regression models. Tidymodels is a popular Machine Learning (ML) library in R that is compatible with the "tidyverse" concepts, and offers various tools for creating and training ML algorithms, feature engineering, data cleaning, and evaluating and testing models. It is the next-gen version of the popular [caret](http://topepo.github.io/caret/index.html) library for R.
```

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+
+In this section, we show you how to apply a simple regression model for predicting tips a server will receive based on various client attributes (such as sex, time of the week, and whether they are a smoker).
+
+We will be using the [Linear Regression][lr], which is a simple model that fit an intercept (the mean tip received by a server), and add a slope for each feature we use, such as the value of the total bill.
```
```suggestion
We will be using the [Linear Regression][lr], which is a simple model that fits an intercept (the mean tip received by a server), and adds a slope for each feature we use, such as the value of the total bill.
```

Note - Instead of `[lr]` I think we want to be consistent with our hyperlinks, I would just make them all Markdown-style.

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+y_range <- lm_model %>% predict(x_range) 
+
+colnames(y_range) <- c('tip')
+xy <- data.frame(x_range, y_range) 
```
One suggestion here is to rename the `dataframe` variable to `df` for this example and other ones (unless there's multiple dataframes).

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
````diff
@@ -0,0 +1,312 @@
+```
+## Model generalization on unseen data
+
+With `add_trace()`, you can easily color your plot based on a predefined data split. By coloring the training and the testing data points with different colors, you can easily see if whether the model generalizes well to the test data or not.
````
```suggestion
With `add_trace()`, you can easily color your plot based on a predefined data split. By coloring the training and the testing data points with different colors, you can easily see if the model generalizes well to the test data or not.
```

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+
+Notice how we can combine scatter points with lines using Plotly. You can learn more about [multiple chart types](https://plotly.com/r/graphing-multiple-chart-types/).
+
+[knn]: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html
```
We'll want to replace this with the R example instead.

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+y <- tips$tip
+X <- tips$total_bill
+
+knn_dist <- nearest_neighbor(neighbors = 10, weight_func = 'inv') %>% 
+  set_engine('kknn') %>% 
+  set_mode('regression') %>%
+  fit(tip ~ total_bill, data = tips)
+knn_uni <- nearest_neighbor(neighbors = 10, weight_func = 'rectangular') %>% 
+  set_engine('kknn') %>% 
+  set_mode('regression') %>%
+  fit(tip ~ total_bill, data = tips) 
```
```suggestion
# Model #1
knn_dist <- nearest_neighbor(neighbors = 10, weight_func = 'inv') %>% 
  set_engine('kknn') %>% 
  set_mode('regression') %>%
  fit(tip ~ total_bill, data = tips)
  
# Model #2
knn_uni <- nearest_neighbor(neighbors = 10, weight_func = 'rectangular') %>% 
  set_engine('kknn') %>% 
  set_mode('regression') %>%
  fit(tip ~ total_bill, data = tips) 
```

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+
+## 3D regression surface with `mesh3d` and `add_surface`
+
+Visualize the decision plane of your model whenever you have more than one variable in your input data. Here, we will use [`svm_rbf`](https://parsnip.tidymodels.org/reference/svm_rbf.html) with [`kernlab`](https://cran.r-project.org/web/packages/kernlab/index.html) engine in `regression` mode. For generating the 2D mesh on the surface, we use the package [`pracma`](https://cran.r-project.org/web/ ...
```
```suggestion
Visualize the decision plane of your model whenever you have more than one variable in your input data. Here, we will use [`svm_rbf`](https://parsnip.tidymodels.org/reference/svm_rbf.html) with [`kernlab`](https://cran.r-project.org/web/packages/kernlab/index.html) engine in `regression` mode. For generating the 2D mesh on the surface, we use the  [`pracma`](https://cran.r-project.org/web/packages/pracma/index.html) package.
```

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+pred <- pred$.pred
+pred <- matrix(pred, dim_val[1], dim_val[2])
+
+dim(pred)
```
Do we need this? If not, I would :scissors:

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+
+scatter <- ggplot(iris, aes(x = Petal.Width, y = prediction, color = split)) +
+  geom_point() + 
+  geom_smooth(method=lm, se=FALSE) 
```
```suggestion
  geom_smooth(formula=y ~ x, method=lm, se=FALSE) 
```

This will avoid the unnecessary warning.

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
```diff
@@ -0,0 +1,312 @@
+
+scatter <- ggplot(iris, aes(x = prediction, y = residual, color = split)) + 
+  geom_point() + 
+  geom_smooth(method=lm, se=FALSE) 
```
```suggestion
  geom_smooth(formula=y ~ x, method=lm, se=FALSE) 
```

HammadTheOne commented on r/2021-07-08-ml-regression.Rmd:
````diff
@@ -0,0 +1,312 @@
+
+```
+## Residual plots
+Just like prediction error plots, it's easy to visualize your prediction residuals in just a few lines of codes using `ggplotly` built-in capabilities.
````
```suggestion
Just like prediction error plots, it's easy to visualize your prediction residuals in just a few lines of codes using `ggplotly` and `tidymodels` capabilities.
```

HammadTheOne reviewed (commented):
Hey @kvdesai , these examples look great! They're fast and fluid, and I like the syntax used. I was able to run all of them by copying and pasting the code into my local environment and running each example script. The `knitted` html also looks good. 

I've made a few suggestions/comments that would bring it across the finish line. Along with that, I would suggest maybe just taking one more pass over the `knitr` generated HTML and addressing any of the non-blocking warnings that appear (eg. the `geom_smooth` formula ones). This will prevent any panic from inexperienced users who might not know what's causing the warning or what they mean. 


Other than that, it looks fantastic. I can approve it once those minor changes have been addressed.

kvdesai commented:
@HammadTheOne - thanks for the review and insightful comments. I have addressed them. I also added one more snippet of the plot. Please verify and let me know if any other change is needed.

HammadTheOne reviewed (approved):
Looks good! Thanks for making the requested changes. 

 I notice that one of the CI tests is failing now though (possibly related to 2016-06-17-3d-tri-surf.Rmd), so once that's fixed we can go ahead and merge it.

:woman_dancing:
